### PR TITLE
Add infer state dim arg

### PR DESCRIPTION
**Pull request by SamDuffield**
Proposed fix for handling factorial case when linearization points are of different shape for init and dynamics. 
Allows the user to provide an optional override for shape inference in `filter_prepare` (default in `build_filter` is to use the init linearization point).

@Sahel13 what do you think? If we go with this I think we'd need to do something similar for `moments`

(Tests all passed locally for me)

### CODE REVIEW

**Comment by Sahel13** [comment]
Wait why do we need this? The fix I suggested in the comment should work no?

**Comment by SamDuffield** [comment]
> Wait why do we need this? The fix I suggested in the comment should work no?

It's a bit subtle/annoying. For factorial inference we ask:

- `init_prepare` to generate mean of shape `(num_factors, x_dim)` and chol_cov of shape `(num_factors, x_dim, x_dim)`
- `filter_prepare` to generate mean of shape `(num_local_factors * x_dim, )` and chol_cov of shape `(num_local_factors * x_dim, num_local_factors * x_dim)`. Since factorial filtering is applied after `join` which flattens into a single dimension.

In the football example we would have `num_factors>20` for all teams in the data and `num_local_factors=2` (matches are between two teams).

So basically we need the option to specify a different latent shape to `filter_prepare` than that of `init_prepare`. Does that make sense?

**Comment by Sahel13** [comment]
> So basically we need the option to specify a different latent shape to filter_prepare than that of init_prepare. Does that make sense?

Yeah I got this, but this is unnecessary? The factorial code we have doesn't support parallel inference, and for the sequential algorithm, the mean and chol_cov saved in the prepared state are ignored. So here we're adding complexity to the generic taylor code so that dummy states for factorial filtering that are discarded have the right shape.

An end-to-end factorial taylor test only fails at the jnp.cov call, and if we replace that, everything works.

**Comment by SamDuffield** [comment]
oh good point! I thought it might complain about shapes but if it works then we're good to go!

**Comment by SamDuffield** [comment]
As discussed, this is not needed!
